### PR TITLE
{devel} Pytorch/0.3.1 (intel and foss 2018a)

### PR DESCRIPTION
**Pull request by wpoely86**
This is based on @zao work in #5530 but with the regular intel and foss toolchains (using CUDA as a versionsuffix).

This needs easybuilders/easybuild-easyblocks#1398 to let it use all the correct dependencies.

### CODE REVIEW

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in this PR)
nic96 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/4dbbb48ebc886a1784d3e42217aee659 for a full test report.

**Comment by hajgato** [comment]
@wpoely86 `PyTorch` tests are very poorly written. They do not take into account what deps were available at installation time.  (if you want all the tests passed)

- GPU shoudl be in `Default` mode, not in `EXCLUSIVE_PROCESS` mode (as set by `Torque`) 
- You need `magma`
- You need `CUDA` aware `OpenMPI` (so will not work with `foss`, only `goolfc` or similar)

**Comment by wpoely86** [comment]
OK, thanks @hajgato 
Magma is merged in #6154 so please add it.

For the tests, can you add some comments in the easyconfig about what you just said?

**Comment by hajgato** [comment]
@wpoley86: I get as far as I can get: it seems that MPI does not work for at all with `PyTorch`, do not know the problem: 
```
RuntimeError: refcounted file mapping not supported on your system at /tmp/hajgato/build/PyTorch/0.3.1/intel-2018a-Python-3.6.4/pytorch-0.3.1/torch/lib/TH/THAllocator.c:525
```
(for both foss and intel, so it seems that the problem is not the CUDA aware MPI (as intel does not have cuda at all)

I was also wrong with the  `EXCLUSIVE_PROCESS` as well.
(`magma` is still needed if you use CUDA)

Using Intel compiler, one test hangs (`test_autograd.py`), when I attach with `strace`, it is in an inifnite loop of:
```
sched_yield()                           = 0
```
I am not able to figure out the cause of this.
What I have until now see:
https://github.com/hajgato/easybuild-easyconfigs/commit/24c031669e3ed746f8ba8a653d1469ff9b88afa4

**Comment by wpoely86** [comment]
That's more then good enough for me. Can I just pull in your branch into mine or are you still working on it?

**Comment by wpoely86** [comment]
@boegel this suffix thing where boegelbot is complaining about, I don't think it's an issue as it's two different toolchains?

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 7 out of 8 (8 easyconfigs in this PR)
nic166 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/28f1d425673407a464a116cacc129fa9 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 8 out of 8 (8 easyconfigs in this PR)
nic96 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/da4ac16a33d45df781609197573c0da2 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 7 out of 8 (8 easyconfigs in this PR)
nic166 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/b3ef53c67b304ee5b051ba81bf0aae56 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 7 out of 8 (8 easyconfigs in this PR)
nic167 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/a21f56917b3a5a288d21660f53e897f9 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
nic167 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/ff59fa1fadf154a26e7a8306903d4fc3 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 8 out of 8 (8 easyconfigs in this PR)
nic167 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/e29fa73477e30106a85ddbc4266fdf2d for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 2 out of 5 (4 easyconfigs in this PR)
rescomp2 - Linux centos 6.8, Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz, Python 2.7.11
See https://gist.github.com/1cd13f67347c5fb6c055732073797ed4 for a full test report.

**Comment by hajgato** [comment]
@wpoely86 I have tried `PyTorch` on magnycours:
`GCC`: version gives illegal instruction in the test pahese (despite it correctly determined the CPU instruction set), so somewhere we have a hard-coded SSSE3+ instruction
`intel`: build of `mkl-dnn` fails all tests. 
So I would not bother to fix that for magnycours.

However, I was able to fix ivybridge-intel, the CPU instruction set detection did not work for `icc`.  Fixed with a patch. (see your pull requests)

**Comment by boegel** [comment]
@wpoely86 I enhanced the tests to allow for the `-CUDA-*` specific variants, see https://github.com/wpoely86/easybuild-easyconfigs/pull/40

**Comment by wpoely86** [comment]
@boegel merging time?

File: easybuild/easyconfigs/p/PyYAML/PyYAML-3.12-foss-2018a-Python-3.6.4.eb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+toolchain = {'name': 'foss', 'version': '2018a'}
+
+source_urls = [PYPI_SOURCE]
+sources = [SOURCE_TAR_GZ]
+checksums = ['592766c6303207a20efc445587778322d7f73b161bd994f227adaa341ba212ab']
+
+dependencies = [
+    ('Python', '3.6.4'),
+    ('libyaml', '0.1.7', '', True),
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyYAML/PyYAML-3.12-foss-2018a-Python-3.6.4.eb]: @wpoely86 Why is this hardcoded to `dummy`? We already have `libyaml-0.1.7-GCCcore-6.4.0.eb` that can resolve this dep.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+PyTorch is a deep learning framework that puts Python first."""
+
+toolchain = {'name': 'foss', 'version': '2018a'}
+
+# This environment is needed to prevent setup.py from invoking git to determine version.
+prebuildopts = 'PYTORCH_BUILD_VERSION=%(version)s PYTORCH_BUILD_NUMBER=1'
+
+# you can choice here: either give a list of CUDA cc version or tell it All
+# by default it does autodetect of the GPU on the local machine
+prebuildopts += ' TORCH_CUDA_ARCH_LIST="3.0 3.2 3.5 3.7 5.0 5.2 5.3 6.0 6.1 7.0"'
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb]: @wpoely86 `prebuildopts` should be moved down, below `dependencies`

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    "https://github.com/facebookincubator/gloo/archive",
+    "https://github.com/google/googletest/archive",
+    "https://github.com/nanopb/nanopb/archive",
+    "https://github.com/pybind/pybind11/archive",
+    "https://github.com/wjakob/clang-cindex-python3/archive",
+]
+sources = [
+    'v%(version)s.tar.gz',  # PyTorch
+    {
+        'filename': 'cb002e4eb8d167c2c60fc3bdaae4e1844e0f9353.tar.gz',
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb]: @wpoely86 Can we use `download_file` here, and rename the downloaded file using `filename` to a more meaningful filename like `gloo-<datestamp-of-commit>.tar.gz` (same below)?

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    # 9f6a636e547fc70a02fa48436449aad67080698f.tar.gz
+    '3a53dac15e87056a9724b9b6403343b47bb71fec26c090979562f71d7ec9ee98',
+    # 6a00cbc4a9b8e68b71caf7f774b3f9c753ae84d5.tar.gz
+    '828e0d6238e2129a9e08071750dc16ba10e38eacf96f21b8a71e501c2085b282',
+    '710bbd38973a94c5c20695d25b819c4b41565bd4723ee4e07542d9537fa2bf18',  # PyTorch-0.3.1-cmake.patch
+    '72c8d6804bff568c2a49b1b7019171c93309a1853a90d5e3abe778f8aa80ca4f',  # PyTorch-0.3.1_skip_MPI_tests.patch
+]
+
+# PyTorch pulls in a bunch of submodules which don't have releases.
+# We download the submodule revisions from their repos.
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb]: @wpoely86 This doesn't belong here, but with `sources`?

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+options = {'modulename': 'torch'}
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb]: @wpoely86 This should be moved down, below `sanity_check_paths`

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+versionsuffix = '-Python-%(pyver)s'
+
+homepage = 'http://pytorch.org/'
+description = """Tensors and Dynamic neural networks in Python with strong GPU acceleration.
+PyTorch is a deep learning framework that puts Python first."""
+
+toolchain = {'name': 'intel', 'version': '2018a'}
+
+# This environment is needed to prevent setup.py from invoking git to determine version.
+prebuildopts = 'PYTORCH_BUILD_VERSION=%(version)s PYTORCH_BUILD_NUMBER=1'
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 please move down

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    "https://github.com/facebookincubator/gloo/archive",
+    "https://github.com/google/googletest/archive",
+    "https://github.com/nanopb/nanopb/archive",
+    "https://github.com/pybind/pybind11/archive",
+    "https://github.com/wjakob/clang-cindex-python3/archive",
+]
+sources = [
+    'v%(version)s.tar.gz',  # PyTorch
+    {
+        'filename': 'cb002e4eb8d167c2c60fc3bdaae4e1844e0f9353.tar.gz',
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 Please use `download_filename` & rename

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    '828e0d6238e2129a9e08071750dc16ba10e38eacf96f21b8a71e501c2085b282',
+    '2b44b3bcda05acbd58c03a6be5a491ad43f58a31a2d0ba845d241a32df47a136',  # PyTorch-0.3.1-fix-for-intel.patch
+    '710bbd38973a94c5c20695d25b819c4b41565bd4723ee4e07542d9537fa2bf18',  # PyTorch-0.3.1-cmake.patch
+    '72c8d6804bff568c2a49b1b7019171c93309a1853a90d5e3abe778f8aa80ca4f',  # PyTorch-0.3.1_skip_MPI_tests.patch
+    '2c07d9c3622986df3ec8d7f761dc6984f1b44c7a377cad7d7667a030aaeb526f',  # PyTorch-0.3.1_intel_skip_autgrad_test.patch
+    '951d5fde89df3be98c1bc6750546259c1b3d0c77574fa3f9e67d9e8fe5b08161',  # PyTorch-0.3.1_fix_findAVX.patch
+]
+
+# PyTorch pulls in a bunch of submodules which don't have releases.
+# We download the submodule revisions from their repos.
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 comment in wrong place

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+options = {'modulename': 'torch'}
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 please move down

File: easybuild/easyconfigs/t/torchvision/torchvision-0.2.0-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+cuda_ver = '9.1.85'
+versionsuffix = '-Python-%%(pyver)s-CUDA-%s' % cuda_ver
+
+homepage = 'https://github.com/pytorch/vision'
+description = " Datasets, Transforms and Models specific to Computer Vision"
+
+toolchain = {'name': 'foss', 'version': '2018a'}
+
+source_urls = [
+    'https://github.com/pytorch/vision/archive',
```
Comment (boegel) [inline easybuild/easyconfigs/t/torchvision/torchvision-0.2.0-foss-2018a-Python-3.6.4-CUDA-9.1.85.eb]: @wpoely86 single line?

File: easybuild/easyconfigs/t/torchvision/torchvision-0.2.0-intel-2018a-Python-3.6.4.eb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+version = '0.2.0'
+versionsuffix = '-Python-%(pyver)s'
+
+homepage = 'https://github.com/pytorch/vision'
+description = " Datasets, Transforms and Models specific to Computer Vision"
+
+toolchain = {'name': 'intel', 'version': '2018a'}
+
+source_urls = [
+    'https://github.com/pytorch/vision/archive',
```
Comment (boegel) [inline easybuild/easyconfigs/t/torchvision/torchvision-0.2.0-intel-2018a-Python-3.6.4.eb]: @wpoely86 single line?

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 7 out of 8 (8 easyconfigs in this PR)
nic167 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/636407f46cc272d57d33fce9e3f49d49 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 5 out of 8 (8 easyconfigs in this PR)
nic44 - Linux centos linux 7.4.1708, AMD Opteron(tm) Processor 6134, Python 2.7.5
See https://gist.github.com/1e1fe532fbe88a4bfba5beaf8579c107 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 8 out of 8 (8 easyconfigs in this PR)
nic151 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/657d8d234f26be1599a568b63594a790 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**FAILED**
Build succeeded for 7 out of 8 (8 easyconfigs in this PR)
nic166 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/c0dd1ece085466daf4e60649950b7838 for a full test report.

**Comment by boegel** [comment]
@wpoely86 what's up with those failing tests?

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 8 out of 8 (8 easyconfigs in this PR)
nic166 - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/fe240b9456a23f64d3027b23c2698128 for a full test report.

**Comment by wpoely86** [comment]
OK, put the GPU mode to shared fixed it. @boegel good to go?

**Comment by boegel** [comment]
Going in, thanks @wpoely86!

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    "https://github.com/nanopb/nanopb/archive",
+    "https://github.com/pybind/pybind11/archive",
+    "https://github.com/wjakob/clang-cindex-python3/archive",
+]
+# PyTorch pulls in a bunch of submodules which don't have releases.
+# We download the submodule revisions from their repos.
+sources = [
+    'v%(version)s.tar.gz',  # PyTorch
+    {
+        'download_filename': 'cb002e4eb8d167c2c60fc3bdaae4e1844e0f9353.tar.gz',
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 @smoors @hajgato @zao How was this determined? Latest & great commit in https://github.com/facebookincubator/gloo ?

Same questions below...

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        'filename': 'gloo-20180105.tar.gz',
+        'extract_cmd': extract_cmd_pattern % (pytorchdir, 'torch/lib/gloo'),
+    },
+    {
+        'download_filename': 'ec44c6c1675c25b9827aacd08c02433cccde7780.tar.gz',
```
Comment (boegel) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: @wpoely86 @smoors @hajgato @zao This commits points to an old version of `googletest`, any specific reason for that?

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb (hunk 12 above)
Comment (wpoely86) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: If I remember correctly they come from the `.gitsubmodules` files.

Comment (boegel) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: That helps, see https://github.com/pytorch/pytorch/blob/master/.gitmodules, but it doesn't explain the old `googletest` commit...

File: easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb (hunk 13 above)
Comment (zao) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: They are exactly the hashes that the submodule indicated when pulling the 0.3.0 tag. I have no idea why they use the version they do.
My workflow was to do a recursive init of submodules and mechanically copy the hashes and paths into my recipe.

Comment (boegel) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-0.3.1-intel-2018a-Python-3.6.4.eb]: It looks like they re-organized things a bit, now you need to look at https://github.com/pytorch/pytorch/tree/v0.4.1/third_party...

**Pull request merged**